src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c

```diff
@@ -0,0 +1,1042 @@
+			}
+
+			mutex_lock(&xgq->xgq_lock);
+
+			if (xgq_consume(&xgq->xgq_queue, &slot_addr)) {
+				mutex_unlock(&xgq->xgq_lock);
+				continue;
+			}
+
+			read_completion(&ccmd, slot_addr);
+
+			cmd_complete(xgq, &ccmd);
+
+			xgq_notify_peer_consumed(&xgq->xgq_queue);
+
+			mutex_unlock(&xgq->xgq_lock);
+		}
+
+		if (xgq->xgq_polling) {
+			usleep_range(1000, 2000);
+		} else {
+			wait_for_completion_interruptible(&xgq->xgq_irq_complete);
+		}
+
+		if (kthread_should_stop()) {
+			xw->stop = true;
+		}
+	}
+	
+	return xw->error ? 1 : 0;
+}
+
+static bool xgq_submitted_cmd_check(struct xocl_xgq *xgq)
+{
+	struct xocl_xgq_cmd *xgq_cmd;
+	struct list_head *pos, *next;
+	bool found_timeout = false;
+
+	mutex_lock(&xgq->xgq_lock);
+	list_for_each_safe(pos, next, &xgq->xgq_submitted_cmds) {
+		xgq_cmd = list_entry(pos, struct xocl_xgq_cmd, xgq_cmd_head);
+
+		XGQ_INFO(xgq, "cmd %llx, jiffies %llx",
+			xgq_cmd->xgq_cmd_timeout_jiffies, jiffies);
+
+		/* Finding timed out cmds */
+		if (xgq_cmd->xgq_cmd_timeout_jiffies < jiffies) {
+			XGQ_ERR(xgq, "cmd id: %d timed out, hot reset is required!",
+				xgq_cmd->xgq_sq.xgq_sq_head.cid);
+			found_timeout = true;
+			break;
+		}
+	}
+	mutex_unlock(&xgq->xgq_lock);
+
+	return found_timeout;
+}
+
+static void xgq_submitted_cmds_drain(struct xocl_xgq *xgq)
+{
+	struct xocl_xgq_cmd *xgq_cmd;
+	struct list_head *pos, *next;
+
+	mutex_lock(&xgq->xgq_lock);
+	list_for_each_safe(pos, next, &xgq->xgq_submitted_cmds) {
+		xgq_cmd = list_entry(pos, struct xocl_xgq_cmd, xgq_cmd_head);
+
+		XGQ_INFO(xgq, "cmd %llx, jiffies %llx",
+			xgq_cmd->xgq_cmd_timeout_jiffies, jiffies);
+
+		/* Finding timed out cmds */
+		if (xgq_cmd->xgq_cmd_timeout_jiffies < jiffies) {
+			list_del(pos);
+			
+			xgq_cmd->xgq_cmd_rcode = -ETIME;
+			complete(&xgq_cmd->xgq_cmd_complete);
+			XGQ_ERR(xgq, "cmd id: %d timed out, hot reset is required!",
+				xgq_cmd->xgq_sq.xgq_sq_head.cid);
+		}
+	}
+	mutex_unlock(&xgq->xgq_lock);
+}
+
+/*
+ * When driver detach, we need to wait for all commands to drain.
+ * If the one command is already timedout, we can safely recycle it only
+ * after disable interrupts and mark device in bad state, a hot_reset
+ * is needed to recover the device back to normal.
+ */
+static bool xgq_submitted_cmds_empty(struct xocl_xgq *xgq)
+{
+	mutex_lock(&xgq->xgq_lock);
+	if (list_empty(&xgq->xgq_submitted_cmds)) {
+		mutex_unlock(&xgq->xgq_lock);
+		return true;
+	}
+	mutex_unlock(&xgq->xgq_lock);
+	
+	return false;
+}
+
+/*
+ * stop service will be called from driver remove or found timeout cmd from health_worker
+ * 3 steps to stop the service:
+ *   1) halt any incoming request
+ *   2) disable interrupts
+ *   3) poll all existing cmds till finish or timeout
+ *
+ * then, we can safely remove all resources.
+ */
+static void xgq_stop_services(struct xocl_xgq *xgq)
+{
+	/* stop receiving incoming commands */
+	mutex_lock(&xgq->xgq_lock);
+	xgq->xgq_halted = 1;
+	mutex_unlock(&xgq->xgq_lock);
+#if 0	
```

Comment:
Hmm...why is this code commented out.   If it is for future debugging, it should have a comment indicating that the code is used for.